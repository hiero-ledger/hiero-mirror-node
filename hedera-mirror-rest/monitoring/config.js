// SPDX-License-Identifier: Apache-2.0

import extend from 'extend';
import fs from 'fs';
import _ from 'lodash';
import logger from './logger';
import path from 'path';
import {fileURLToPath} from 'url';

const REQUIRED_FIELDS = [
  'servers',
  'interval',
  'shard',
  'timeout',
  'account.intervalMultiplier',
  'balance.freshnessThreshold',
  'balance.intervalMultiplier',
  'block.freshnessThreshold',
  'block.intervalMultiplier',
  'network.intervalMultiplier',
  'stateproof.intervalMultiplier',
  'transaction.freshnessThreshold',
  'transaction.intervalMultiplier',
  'topic.freshnessThreshold',
  'topic.intervalMultiplier',
];

const load = (configFile) => {
  try {
    if (!configFile) {
      return {};
    }

    const data = JSON.parse(fs.readFileSync(configFile).toString('utf8'));
    logger.info(`Loaded configuration source: ${configFile}`);
    return data;
  } catch (err) {
    logger.warn(`Skipping configuration source ${configFile}: ${err}`);
    return {};
  }
};

let config = {};
let loaded = false;

if (!loaded) {
  const moduleDirname = path.dirname(fileURLToPath(import.meta.url));
  config = load(path.join(moduleDirname, 'config', 'default.serverlist.json'));
  const customConfig = load(path.join(moduleDirname, 'config', 'serverlist.json'));
  extend(true, config, customConfig);

  if (process.env.CONFIG_PATH) {
    const customPathConfig = load(path.join(process.env.CONFIG_PATH, 'serverlist.json'));
    extend(true, config, customPathConfig);
  }

  for (const field of REQUIRED_FIELDS) {
    if (!_.has(config, field)) {
      throw new Error(`required field "${field}" not found in any configuration file`);
    }
  }

  if (!Array.isArray(config.servers) || config.servers.length === 0) {
    throw new Error(`Invalid servers "${JSON.stringify(config.servers)}" in any configuration file`);
  }

  logger.info(`Loaded configuration: ${JSON.stringify(config)}`);
  loaded = true;
}

export default config;
