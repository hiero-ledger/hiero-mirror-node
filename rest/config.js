// SPDX-License-Identifier: Apache-2.0

import extend from 'extend';
import fs from 'fs';
import yaml from 'js-yaml';
import parseDuration from 'parse-duration';
import path from 'path';
import {fileURLToPath} from 'url';

import {cloudProviders, defaultBucketNames, NANOSECONDS_PER_MILLISECOND, networks} from './constants';
import {InvalidConfigError} from './errors';
import configureLogger from './logger';

configureLogger();

const config = {};
const defaultConfigName = 'application';
const hederaPrefix = 'hedera';
const hieroPrefix = 'hiero';
let loaded = false;

function load(configPath, configName) {
  if (!configPath) {
    return;
  }

  let configFile = path.join(configPath, `${configName}.yml`);
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }

  configFile = path.join(configPath, `${configName}.yaml`);
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }
}

function loadYaml(configFile) {
  try {
    const doc = yaml.load(fs.readFileSync(configFile, 'utf8'));
    logger.info(`Loaded configuration source: ${configFile}`);
    extend(true, config, doc);

    // Migrated deprecated properties
    const hedera = doc[hederaPrefix];
    if (hedera) {
      extend(true, config, {hiero: hedera});
      delete config[hederaPrefix];
      logger.warn(
        `Source contains deprecated '${hederaPrefix}' properties that have been automatically migrated to '${hieroPrefix}''`
      );
    }
  } catch (err) {
    logger.warn(`Skipping configuration ${configFile}: ${err}`);
  }
}

function loadEnvironment() {
  for (const [key, value] of Object.entries(process.env)) {
    setConfigValue(key, value);
  }
}

/*
 * Sets a config property from an environment variable by converting HIERO_MIRROR_REST_FOO_BAR to an object path
 * notation hiero.mirror.rest.foo.bar using a case insensitive search. If more than one property matches with a
 * different case, it will choose the first. Will also convert HEDERA_MIRROR to hiero.mirror properties.
 */
function setConfigValue(propertyPath, value) {
  let current = config;
  const properties = propertyPath.toLowerCase().split('_');

  // Ignore properties that don't start with HEDERA_MIRROR or HIERO_MIRROR
  if (
    properties.length < 3 ||
    (properties[0] !== 'hedera' && properties[0] !== 'hiero') ||
    properties[1] !== 'mirror'
  ) {
    return;
  }

  for (let i = 0; i < properties.length; i += 1) {
    let property = properties[i];
    let found = false;

    for (const [k, v] of Object.entries(current)) {
      if (property === hederaPrefix) {
        property = hieroPrefix;
        logger.warn(
          `Deprecated '${hederaPrefix}' property automatically migrated to '${hieroPrefix}': ${propertyPath}`
        );
      }

      if (property === k.toLowerCase()) {
        if (i < properties.length - 1) {
          current = v;
          found = true;
          break;
        } else {
          current[k] = convertType(value);
          const cleanedValue = property.includes('password') || property.includes('key') ? '******' : value;
          logger.info(`Override config with environment variable ${propertyPath}=${cleanedValue}`);
          return;
        }
      }
    }

    if (!found) {
      return;
    }
  }
}

function convertType(value) {
  if (value !== null && value !== '' && !isNaN(value)) {
    return +value;
  } else if (value === 'true' || value === 'false') {
    return value === 'true';
  }

  return value;
}

function getConfig() {
  return config.hiero?.mirror?.rest;
}

function getMirrorConfig() {
  return config.hiero?.mirror;
}

function getResponseLimit() {
  return getConfig().response.limit;
}

function parseDbPoolConfig() {
  const {pool} = getConfig().db;
  const configKeys = ['connectionTimeout', 'maxConnections', 'statementTimeout'];
  configKeys.forEach((configKey) => {
    const value = pool[configKey];
    const parsed = parseInt(value, 10);
    if (Number.isNaN(parsed) || parsed <= 0) {
      throw new InvalidConfigError(`invalid value set for db.pool.${configKey}: ${value}`);
    }
    pool[configKey] = parsed;
  });
}

function parseStateProofStreamsConfig() {
  const {stateproof} = getConfig();
  if (!stateproof || !stateproof.enabled) {
    return;
  }

  const {streams: streamsConfig} = stateproof;
  if (!Object.values(networks).includes(streamsConfig.network)) {
    throw new InvalidConfigError(`unknown network ${streamsConfig.network}`);
  }

  if (!streamsConfig.bucketName) {
    streamsConfig.bucketName = defaultBucketNames[streamsConfig.network];
  }

  if (!streamsConfig.bucketName) {
    // the default for network 'OTHER' is null, throw err if it's not configured
    throw new InvalidConfigError('stateproof.streams.bucketName must be set');
  }

  if (!Object.values(cloudProviders).includes(streamsConfig.cloudProvider)) {
    throw new InvalidConfigError(`unsupported object storage service provider ${streamsConfig.cloudProvider}`);
  }
}

const parseDurationConfig = (name, value) => {
  const ms = parseDuration(value);
  if (!ms) {
    throw new InvalidConfigError(`invalid ${name} ${value}`);
  }
  return BigInt(ms) * NANOSECONDS_PER_MILLISECOND;
};

const durationQueryConfigKeys = [
  'maxRecordFileCloseInterval',
  'maxScheduledTransactionConsensusTimestampRange',
  'maxTimestampRange',
  'maxTransactionConsensusTimestampRange',
  'maxTransactionsTimestampRange',
];

const parseQueryConfig = () => {
  const {query} = getConfig();
  const {precedingTransactionTypes} = query.transactions;
  if (!Array.isArray(precedingTransactionTypes)) {
    throw new InvalidConfigError(
      `Invalid or missing query.transactions.precedingTransactionTypes: ${precedingTransactionTypes}`
    );
  }
  durationQueryConfigKeys.forEach((key) => (query[`${key}Ns`] = parseDurationConfig(`query.${key}`, query[key])));
};

const parseNetworkConfig = () => {
  const currencyFormat = getConfig().network.currencyFormat;
  if (currencyFormat) {
    const validValues = ['BOTH', 'HBARS', 'TINYBARS'];
    if (!validValues.includes(currencyFormat)) {
      throw new InvalidConfigError(`invalid currencyFormat ${currencyFormat}`);
    }
  }
};

const parseCommon = () => {
  const {common} = getMirrorConfig();
  if (common?.shard !== undefined) {
    common.shard = BigInt(common.shard);
  }
  if (common?.realm !== undefined) {
    common.realm = BigInt(common.realm);
  }
};

if (!loaded) {
  const configName = process.env.CONFIG_NAME || defaultConfigName;
  // always load the default configuration
  const moduleDirname = path.dirname(fileURLToPath(import.meta.url));
  load(path.join(moduleDirname, 'config'), defaultConfigName);
  load(moduleDirname, configName);
  load(process.env.CONFIG_PATH, configName);
  loadEnvironment();
  parseDbPoolConfig();
  parseNetworkConfig();
  parseQueryConfig();
  parseStateProofStreamsConfig();
  parseCommon();
  loaded = true;
  configureLogger(getConfig().log.level);
}

export default getConfig();

export {getMirrorConfig, getResponseLimit};
