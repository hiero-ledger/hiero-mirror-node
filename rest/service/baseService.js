// SPDX-License-Identifier: Apache-2.0

import _ from 'lodash';

/**
 * Base service class that other services should inherit from for their retrieval business logic
 */
class BaseService {
  buildWhereSqlStatement(whereQuery, params = []) {
    if (_.isEmpty(whereQuery)) {
      return {where: '', params};
    }

    const prefix = params.length === 0 ? 'where' : 'and';
    const condition = whereQuery
      .map((current) => {
        let position;
        if (Array.isArray(current.param)) {
          const first = params.length + 1;
          const last = params.push(...current.param);
          position = `(${_.range(first, last + 1).map((pos) => '$' + pos)})`;
        } else {
          params.push(current.param);
          position = `$${params.length}`;
        }
        return `${current.query} ${position}`;
      })
      .join(' and ');

    return {where: `${prefix} ${condition}`, params};
  }

  getLimitQuery(position) {
    return `limit $${position}`;
  }

  /**
   * Gets the order by query from the order specs
   *
   * @param {OrderSpec} orderSpecs
   * @return {string}
   */
  getOrderByQuery(...orderSpecs) {
    return 'order by ' + orderSpecs.map((spec) => `${spec}`).join(', ');
  }

  getFilterWhereCondition(key, filter) {
    return {
      query: `${key} ${filter.operator}`,
      param: filter.value,
    };
  }

  async getRows(query, params) {
    return (await this.pool().queryQuietly(query, params)).rows;
  }

  async getSingleRow(query, params) {
    const rows = await this.getRows(query, params);
    if (_.isEmpty(rows) || rows.length > 1) {
      return null;
    }

    return rows[0];
  }

  /**
   * Builds a standard sql query that can be extended with an additional filters
   *
   * @param {string} selectFromTable
   * @param {*[]} params
   * @param {string[]} conditions
   * @param {string} orderClause
   * @param {string} limitClause
   * @param {{key: string, operator: string, value: *, column: string}[]} filters
   * @returns {(string|*)[]} the build query and the parameters for the query
   */
  buildSelectQuery(selectFromTable, params, conditions, orderClause, limitClause, filters = []) {
    const whereConditions = [
      ...conditions,
      ...filters.map((filter) => {
        params.push(filter.value);
        return `${filter.column}${filter.operator}$${params.length}`;
      }),
    ];

    return [
      selectFromTable,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      orderClause,
      limitClause,
    ].join('\n');
  }

  pool() {
    return pool;
  }
}

export default BaseService;
