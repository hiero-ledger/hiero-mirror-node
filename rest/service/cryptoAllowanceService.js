// SPDX-License-Identifier: Apache-2.0

import BaseService from './baseService';
import {CryptoAllowance} from '../model';
import {OrderSpec} from '../sql';

/**
 * CryptoAllowance business model
 */
class CryptoAllowanceService extends BaseService {
  static accountAllowanceQuery = `select * from ${CryptoAllowance.tableName}`;

  async getAccountCryptoAllowances(conditions, initParams, order, limit) {
    const {query, params} = this.getAccountAllowancesQuery(conditions, initParams, order, limit);
    const rows = await super.getRows(query, params);
    return rows.map((ca) => new CryptoAllowance(ca));
  }

  getAccountAllowancesQuery(whereConditions, whereParams, order, limit) {
    const params = whereParams;
    params.push(limit);
    const query = [
      CryptoAllowanceService.accountAllowanceQuery,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      super.getOrderByQuery(OrderSpec.from(CryptoAllowance.SPENDER, order)),
      super.getLimitQuery(params.length),
    ].join('\n');

    return {query, params};
  }
}

export default new CryptoAllowanceService();
