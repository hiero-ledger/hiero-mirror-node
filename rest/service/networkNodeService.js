// SPDX-License-Identifier: Apache-2.0

import BaseService from './baseService';
import {AddressBook, AddressBookEntry, AddressBookServiceEndpoint, Node, NetworkNode, NodeStake} from '../model';
import {OrderSpec} from '../sql';
import EntityId from '../entityId';

/**
 * Network node business model
 */
class NetworkNodeService extends BaseService {
  // add node filter
  static networkNodesBaseQuery = `with ${AddressBook.tableAlias} as (
      select ${AddressBook.START_CONSENSUS_TIMESTAMP}, ${AddressBook.END_CONSENSUS_TIMESTAMP}, ${AddressBook.FILE_ID}
      from ${AddressBook.tableName} where ${AddressBook.FILE_ID} = $1
      order by ${AddressBook.START_CONSENSUS_TIMESTAMP} desc limit 1
    ),
    ${NodeStake.tableAlias} as (
      select ${NodeStake.MAX_STAKE}, ${NodeStake.MIN_STAKE}, ${NodeStake.NODE_ID}, ${NodeStake.REWARD_RATE},
             ${NodeStake.STAKE}, ${NodeStake.STAKE_NOT_REWARDED}, ${NodeStake.STAKE_REWARDED},
             ${NodeStake.STAKING_PERIOD}
      from ${NodeStake.tableName}
      where ${NodeStake.CONSENSUS_TIMESTAMP} =
        (select max(${NodeStake.CONSENSUS_TIMESTAMP}) from ${NodeStake.tableName})
    ),
    ${Node.tableAlias} as (
      select ${Node.ADMIN_KEY}, ${Node.DECLINE_REWARD}, ${Node.GRPC_PROXY_ENDPOINT}, ${Node.NODE_ID}
      from ${Node.tableName}
    )
    select ${AddressBookEntry.getFullName(AddressBookEntry.DESCRIPTION)},
      ${AddressBookEntry.getFullName(AddressBookEntry.MEMO)},
      ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)},
      ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ACCOUNT_ID)},
      ${AddressBookEntry.getFullName(AddressBookEntry.NODE_CERT_HASH)},
      ${AddressBookEntry.getFullName(AddressBookEntry.PUBLIC_KEY)},
      ${AddressBook.getFullName(AddressBook.FILE_ID)},
      ${AddressBook.getFullName(AddressBook.START_CONSENSUS_TIMESTAMP)},
      ${AddressBook.getFullName(AddressBook.END_CONSENSUS_TIMESTAMP)},
      ${Node.getFullName(Node.ADMIN_KEY)},
      ${Node.getFullName(Node.DECLINE_REWARD)},
      ${Node.getFullName(Node.GRPC_PROXY_ENDPOINT)},
      ${NodeStake.getFullName(NodeStake.MAX_STAKE)},
      ${NodeStake.getFullName(NodeStake.MIN_STAKE)},
      ${NodeStake.getFullName(NodeStake.REWARD_RATE)},
      ${NodeStake.getFullName(NodeStake.STAKE)},
      ${NodeStake.getFullName(NodeStake.STAKE_NOT_REWARDED)},
      ${NodeStake.getFullName(NodeStake.STAKE_REWARDED)},
      ${NodeStake.getFullName(NodeStake.STAKING_PERIOD)},
      coalesce((
        select jsonb_agg(jsonb_build_object(
        '${AddressBookServiceEndpoint.IP_ADDRESS_V4}', ${AddressBookServiceEndpoint.IP_ADDRESS_V4},
        '${AddressBookServiceEndpoint.PORT}', ${AddressBookServiceEndpoint.PORT},
        '${AddressBookServiceEndpoint.DOMAIN_NAME}', ${AddressBookServiceEndpoint.DOMAIN_NAME}
        ) order by ${AddressBookServiceEndpoint.IP_ADDRESS_V4} asc, ${AddressBookServiceEndpoint.PORT} asc)
        from ${AddressBookServiceEndpoint.tableName} ${AddressBookServiceEndpoint.tableAlias}
        where ${AddressBookServiceEndpoint.getFullName(AddressBookServiceEndpoint.CONSENSUS_TIMESTAMP)} =
          ${AddressBookEntry.getFullName(AddressBookEntry.CONSENSUS_TIMESTAMP)} and
          ${AddressBookServiceEndpoint.getFullName(AddressBookServiceEndpoint.NODE_ID)} =
          ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)}
      ), '[]') as service_endpoints
    from ${AddressBookEntry.tableName} ${AddressBookEntry.tableAlias}
    join ${AddressBook.tableAlias} on ${AddressBook.getFullName(AddressBook.START_CONSENSUS_TIMESTAMP)} =
      ${AddressBookEntry.getFullName(AddressBookEntry.CONSENSUS_TIMESTAMP)}
    left join ${NodeStake.tableAlias} on ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)} =
      ${NodeStake.getFullName(NodeStake.NODE_ID)}
    left join ${Node.tableAlias} on ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)} =
      ${Node.getFullName(Node.NODE_ID)}`;

  static unreleasedSupplyAccounts = (column) =>
    EntityId.systemEntity.unreleasedSupplyAccounts
      .map((range) => {
        const from = range.from.getEncodedId();
        const to = range.to.getEncodedId();

        if (from === to) {
          return `${column} = ${from}`;
        } else {
          return `(${column} >= ${from} and ${column} <= ${to})`;
        }
      })
      .join(' or ');

  static networkSupplyQuery = `
      select coalesce(sum(balance), 0) as unreleased_supply, coalesce(max(balance_timestamp), 0) as consensus_timestamp
      from entity
      where ${NetworkNodeService.unreleasedSupplyAccounts('id')}`;

  getNetworkSupplyByTimestampQuery = (conditions) => `
      with account_balances as (
          select distinct on (account_id) balance, consensus_timestamp
          from account_balance ab
          where ${conditions} and (${NetworkNodeService.unreleasedSupplyAccounts('account_id')})
          order by account_id asc, consensus_timestamp desc
      )
      select coalesce(sum(balance), 0) as unreleased_supply, max(consensus_timestamp) as consensus_timestamp
      from account_balances`;

  getNetworkNodes = async (whereConditions, whereParams, order, limit) => {
    const [query, params] = this.getNetworkNodesWithFiltersQuery(whereConditions, whereParams, order, limit);

    const rows = await super.getRows(query, params);
    return rows.map((x) => new NetworkNode(x));
  };

  getNetworkNodesWithFiltersQuery = (whereConditions, whereParams, nodeOrder, limit) => {
    const params = whereParams;
    params.push(limit);
    const query = [
      NetworkNodeService.networkNodesBaseQuery,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      `${super.getOrderByQuery(OrderSpec.from(AddressBookEntry.getFullName(AddressBookEntry.NODE_ID), nodeOrder))}`,
      super.getLimitQuery(params.length),
    ].join('\n');

    return [query, params];
  };

  getSupply = async (conditions, params) => {
    let query = NetworkNodeService.networkSupplyQuery;

    if (conditions.length > 0) {
      query = this.getNetworkSupplyByTimestampQuery(conditions.join(' and '));
    }

    return await super.getSingleRow(query, params);
  };
}

export default new NetworkNodeService();
