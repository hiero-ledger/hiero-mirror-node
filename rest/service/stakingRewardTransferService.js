// SPDX-License-Identifier: Apache-2.0

import BaseService from './baseService';
import {StakingRewardTransfer} from '../model';
import {OrderSpec} from '../sql';

/**
 * Staking Reward Transfer retrieval business logic
 */
class StakingRewardTransferService extends BaseService {
  static listStakingRewardsByAccountIdQuery = `
    select ${StakingRewardTransfer.getFullName(StakingRewardTransfer.ACCOUNT_ID)},
    ${StakingRewardTransfer.getFullName(StakingRewardTransfer.AMOUNT)},
    ${StakingRewardTransfer.getFullName(StakingRewardTransfer.CONSENSUS_TIMESTAMP)}
    from ${StakingRewardTransfer.tableName} ${StakingRewardTransfer.tableAlias}
    where ${StakingRewardTransfer.getFullName(StakingRewardTransfer.ACCOUNT_ID)} = $1`;

  async getRewards(order, limit, conditions, initParams) {
    const {query, params} = this.getRewardsQuery(order, limit, conditions, initParams);
    const rows = await super.getRows(query, params);
    return rows.map((srt) => new StakingRewardTransfer(srt));
  }

  getRewardsQuery(order, limit, conditions, params) {
    const query = [
      StakingRewardTransferService.listStakingRewardsByAccountIdQuery,
      conditions.length > 0 ? `and ${conditions.join(' and ')}` : '', // "and" since we already have "where account_id = $1" at the end of the above line
      super.getOrderByQuery(
        OrderSpec.from(StakingRewardTransfer.getFullName(StakingRewardTransfer.CONSENSUS_TIMESTAMP), order)
      ),
      super.getLimitQuery(2), // limit is specified in $2 (not necessarily a limit *of* 2)
    ].join('\n');

    return {query, params};
  }
}

export default new StakingRewardTransferService();
