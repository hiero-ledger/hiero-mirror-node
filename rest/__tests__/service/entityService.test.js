// SPDX-License-Identifier: Apache-2.0

import EntityId from '../../entityId';
import {EntityService} from '../../service';
import AccountAlias from '../../accountAlias';
import integrationDomainOps from '../integrationDomainOps';
import {setupIntegrationTest} from '../integrationUtils';
import {getMirrorConfig} from '../../config.js';

setupIntegrationTest();

const {
  common: {realm: systemRealm, shard: systemShard},
} = getMirrorConfig();

const defaultEntityAlias = new AccountAlias(
  systemShard,
  systemRealm,
  'KGNABD5L3ZGSRVUCSPDR7TONZSRY3D5OMEBKQMVTD2AC6JL72HMQ'
);
const defaultInputEntityId = EntityId.parseString('95622');
const defaultInputEntity = [
  {
    alias: defaultEntityAlias.base32Alias,
    evm_address: 'ac384c53f03855fa1b3616052f8ba32c6c2a2fec',
    id: defaultInputEntityId.getEncodedId(),
    ...defaultInputEntityId,
  },
];
const defaultInputContractId = EntityId.parseString('95623');
const defaultInputContract = [
  {
    evm_address: 'cef2a2c6c23ab8f2506163b1af55830f35c483ca',
    id: defaultInputContractId.getEncodedId(),
    num: defaultInputContractId.num,
    shard: systemShard,
    realm: systemRealm,
  },
];

const defaultExpectedEntity = {id: EntityId.parseString('95622').getEncodedId()};
const defaultExpectedContractId = {id: EntityId.parseString('95623').getEncodedId()};

describe('EntityService.getAccountFromAlias tests', () => {
  test('EntityService.getAccountFromAlias - No match', async () => {
    await expect(EntityService.getAccountFromAlias({alias: '1'})).resolves.toBeNull();
  });

  test('EntityService.getAccountFromAlias - Matching entity', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getAccountFromAlias(defaultEntityAlias)).resolves.toMatchObject(defaultExpectedEntity);
  });

  test('EntityService.getAccountFromAlias - Duplicate alias', async () => {
    const entityId1 = EntityId.parseString('3');
    const entityId2 = EntityId.parseString('4');

    const inputEntities = [
      {
        alias: defaultEntityAlias.base32Alias,
        id: entityId1.getEncodedId(),
        ...entityId1,
      },
      {
        alias: defaultEntityAlias.base32Alias,
        id: entityId2.getEncodedId(),
        ...entityId2,
      },
    ];
    await integrationDomainOps.loadEntities(inputEntities);

    await expect(() => EntityService.getAccountFromAlias(defaultEntityAlias)).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('EntityService.getAccountIdFromAlias tests', () => {
  test('EntityService.getAccountIdFromAlias - No match - result required', async () => {
    await expect(() => EntityService.getAccountIdFromAlias(defaultEntityAlias)).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getAccountIdFromAlias - No match - result not required', async () => {
    await expect(EntityService.getAccountIdFromAlias(defaultEntityAlias, false)).resolves.toBe(null);
  });

  test('EntityService.getAccountFromAlias - Matching id', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getAccountIdFromAlias(defaultEntityAlias)).resolves.toBe(defaultExpectedEntity.id);
  });
});

describe('EntityService.getEntityIdFromEvmAddress tests', () => {
  const entity = defaultInputEntity[0];
  const defaultEvmAddress = EntityId.parse(`${entity.shard}.${entity.realm}.${entity.evm_address}`);

  test('EntityService.getEntityIdFromEvmAddress - Matching evm address', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)).resolves.toBe(defaultExpectedEntity.id);
  });

  test('EntityService.getEntityIdFromEvmAddress - No match - result required', async () => {
    await expect(() =>
      EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getEntityIdFromEvmAddress - No match - result not required', async () => {
    await expect(EntityService.getEntityIdFromEvmAddress(defaultEvmAddress, false)).resolves.toBe(null);
  });

  test('EntityService.getEntityIdFromEvmAddress - Multiple matches', async () => {
    const inputEntities = [
      defaultInputEntity[0],
      {
        ...defaultInputEntity[0],
        id: BigInt(defaultInputEntity[0].id) + 1n,
        num: BigInt(defaultInputEntity[0].num) + 1n,
      },
    ];
    await integrationDomainOps.loadEntities(inputEntities);

    await expect(() =>
      EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getEntityIdFromEvmAddress - Non-zero realm', async () => {
    const entity = {
      evm_address: '81eaa748d5252be68c1185588beca495459fdba4',
      id: 18014948265296876n,
      num: 1004,
      shard: 1,
      realm: 2,
    };
    await integrationDomainOps.loadEntities([entity]);

    await expect(EntityService.getEntityIdFromEvmAddress(EntityId.parse(entity.evm_address))).resolves.toBe(
      18014948265296876n
    );
    await expect(EntityService.getEntityIdFromEvmAddress(EntityId.parse(`0x${entity.evm_address}`))).resolves.toBe(
      18014948265296876n
    );
  });
});

describe('EntityService.isValidAccount tests', () => {
  test('EntityService.isValidAccount - No match', async () => {
    await expect(EntityService.isValidAccount(defaultInputEntity[0].id)).resolves.toBe(false);
  });

  test('EntityService.getAccountFromAlias - Matching', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.isValidAccount(defaultInputEntity[0].id)).resolves.toBe(true);
  });
});

describe('EntityService.getEncodedId tests', () => {
  test('EntityService.getEncodedId - No match', async () => {
    await expect(EntityService.getEncodedId(defaultInputEntity[0].num + '')).resolves.toBe(defaultExpectedEntity.id);
  });

  test('EntityService.getEncodedId - Matching id', async () => {
    await expect(EntityService.getEncodedId(defaultInputEntity[0].num + '')).resolves.toBe(defaultExpectedEntity.id);
    await expect(EntityService.getEncodedId(defaultInputContract[0].num + '')).resolves.toBe(
      defaultExpectedContractId.id
    );
  });

  test('EntityService.getEncodedId - Matching alias', async () => {
    const entityId = EntityId.parseString('1001');
    const entity = {
      alias: 'AAAQEAYEAUDAOCAJCAIREEYUCULBOGAZ',
      id: entityId.getEncodedId(),
      ...entityId,
    };
    await integrationDomainOps.loadEntities([entity]);

    await expect(EntityService.getEncodedId(entity.alias)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`${entity.realm}.${entity.alias}`)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`${entity.shard}.${entity.realm}.${entity.alias}`)).resolves.toBe(
      entity.id
    );
  });

  test('EntityService.getEncodedId - Matching evm address', async () => {
    const entityId = EntityId.parseString('1002');
    const entity = {
      evm_address: '71eaa748d5252be68c1185588beca495459fdba4',
      id: entityId.getEncodedId(),
      ...entityId,
    };
    await integrationDomainOps.loadEntities([entity]);

    await expect(EntityService.getEncodedId(entity.evm_address)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`0x${entity.evm_address}`)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`${entity.realm}.${entity.evm_address}`)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`${entity.shard}.${entity.realm}.${entity.evm_address}`)).resolves.toBe(
      entity.id
    );
  });

  test('EntityService.getEncodedId - Non-default shard/realm', async () => {
    const entityId = EntityId.parseString('1003');
    const entity = {
      alias: 'AEBAGBAFAYDQQCIQCEJBGFAVCYLRQGJA',
      evm_address: '81eaa748d5252be68c1185588beca495459fdba4',
      id: entityId.getEncodedId(),
      ...entityId,
    };
    await integrationDomainOps.loadEntities([entity]);

    await expect(EntityService.getEncodedId(entity.alias)).resolves.toBe(entity.id);
    await expect(
      EntityService.getEncodedId(`${entity.realm + 1}.${entity.alias}`)
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      EntityService.getEncodedId(`${entity.shard + 1}.${entity.realm + 1}.${entity.alias}`)
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(EntityService.getEncodedId(entity.evm_address)).resolves.toBe(entity.id);
    await expect(EntityService.getEncodedId(`0x${entity.evm_address}`)).resolves.toBe(entity.id);
    await expect(
      EntityService.getEncodedId(`${entity.realm + 1}.${entity.evm_address}`)
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      EntityService.getEncodedId(`${entity.shard + 1}.${entity.realm + 1}.${entity.evm_address}`)
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getEncodedId - Invalid alias', async () => {
    await expect(EntityService.getEncodedId('deadbeef=')).rejects.toThrowErrorMatchingSnapshot();
  });
});
