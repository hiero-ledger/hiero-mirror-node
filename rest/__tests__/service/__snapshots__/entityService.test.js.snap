// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EntityService.getAccountFromAlias tests EntityService.getAccountFromAlias - Duplicate alias 1`] = `"Multiple alive entities matching alias"`;

exports[`EntityService.getAccountIdFromAlias tests EntityService.getAccountIdFromAlias - No match - result required 1`] = `"No account with a matching alias found"`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Invalid alias 1`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Non-default shard/realm 1`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Non-default shard/realm 2`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Non-default shard/realm 3`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Non-default shard/realm 4`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEntityIdFromEvmAddress tests EntityService.getEntityIdFromEvmAddress - Multiple matches 1`] = `"Multiple alive entities matching evm address"`;

exports[`EntityService.getEntityIdFromEvmAddress tests EntityService.getEntityIdFromEvmAddress - No match - result required 1`] = `"Not found"`;
